completions/tipc

```diff
@@ -0,0 +1,311 @@
+# tipc(8) completion                                       -*- shell-script -*-
+
+# based on key in $1, get the value from $words[]
+_tipc_get_val() {
+    kval=""
+    for ((i = 0; i < $cword; i++)); do
+        if [ ${words[$i]} == $1 ]; then
+            kval=${words[$(($i + 1))]}
+            return 1
+        fi
+    done
+    return 0
+}
+
+_tipc_media() {
+    local optind=$1
+
+    if [[ $cword -eq $optind ]]; then
+        COMPREPLY=( $( compgen -W 'media' -- $cur) )
+        return 0
+    elif [[ $cword -eq $optind+1 ]]; then
+        COMPREPLY=( $( compgen -W 'udp eth ib' -- $cur) )
+        return 0
+    fi
+
+    return 1
+}
+
+_tipc_bearer() {
+    local optind=$1
+    local kval media
+
+    if _tipc_media $optind; then
+        return 1
+    fi
+
+    _tipc_get_val 'media'
+    media="$kval"
+
+    if [[ $cword -eq $optind+2 ]]; then
+        case "$media" in
+        "udp")
+            COMPREPLY=( $( compgen -W 'name' -- $cur) )
+            ;;
+        "eth" | "ib")
+            COMPREPLY=( $( compgen -W 'device' -- $cur) )
+            ;;
+        esac
+    elif [[ $cword -eq $optind+3 ]]; then
+        case "$media" in
+        "udp")
+            local names=$(tipc bearer list | awk -F: '/^udp:/ {print $2}')
+            COMPREPLY=( $( compgen -W '$names' -- $cur) )
+            ;;
+        "eth")
+            local interfaces=$(ls /sys/class/net/)
+            COMPREPLY=( $( compgen -W '$interfaces' -- $cur ) )
+            ;;
+        esac
+    fi
+}
+
+_tipc_link_opts() {
+    COMPREPLY=( $( compgen -W 'priority tolerance window' -- $cur) )
+}
+
+_tipc_link() {
+    local optind=$1
+    local filter=$2
+
+    if [[ $cword -eq $optind ]]; then
+        COMPREPLY=( $( compgen -W 'link' -- $cur) )
+    elif [[ $cword -eq $optind+1 ]]; then
+        # awk drops link state and last trailing :
+        local links=$(tipc link list | awk '{print substr($1, 0, length($1)-1)}')
+        if [[ $filter == "peers" ]]; then
+            links=$(echo "$links" | sed '/broadcast-link/d')
+        fi
+        COMPREPLY=( $( compgen -W '$links' -- $cur ) )
+    fi
+}
+
+_tipc()
+{
+    local cur prev words cword optind
+    _init_completion || return
+
+    optind=1
+    COMPREPLY=()
+
+    # Flags can be placed anywhere in the commandline
+    case "${words[$cword]}" in
```

Comment:
It is when completing the first sub-command. I.e. "tipc bearer" will switch correctly. When an additional command is passed "tipc bearer enable" $cur will be "enable", making this switch fail. I need to travers down the command tree looking at each passed subcommand. I didn't find any cleaner way to do that.